File: Competition/2021/Golden/Single-Digit/Solution.java
// Intercollegiate Programming Competition 2021 Gold Track
// Problem 2 Single Digit

import java.util.Scanner;
import java.util.ArrayList;
import java.lang.Math;

public class Solution
{
    public static void main(String args[])
    {

        // Get input from STDIN
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter a number: ");

        // Get number as a string
        String numStr = scanner.nextLine();

        scanner.close();

        // Get the list of digits from the number
        ArrayList<Integer> digits = splitToDigits(numStr);

        // List to be used in the while loop to find final single digit
        ArrayList<Integer> resultingList = new ArrayList<Integer>();

        for (int digit : digits)
        {
            resultingList.add(digit);
        }

        // Variable that controls if the final result is one single digit
        boolean oneDigit = resultingList.size() == 1;

        // As long as the final result is not single digit, keep subtracting
        while (!(oneDigit))
        {
            resultingList = getSubDigits(resultingList);

            if (onlyOne(resultingList).size() == 1 || resultingList.size() == 1)
            {
                oneDigit = true;
            }
        }

        if (onlyOne(resultingList).size() == 1)
        {
            System.out.println(onlyOne(resultingList).get(0));
        }

        else
        {
            System.out.println(resultingList.get(0));
        }

    }

    public static ArrayList<Integer> splitToDigits(String num)
    {

        // Final Array List of Digits
        ArrayList<Integer> digits = new ArrayList<Integer>();

        // Loop through every digit of the number
        for (int i = 0; i < num.length(); i++)
        {
            int digit = Integer.parseInt(String.valueOf(num.charAt(i)));
            digits.add(digit);
        }

        return digits;

    }

    public static ArrayList<Integer> onlyOne(ArrayList<Integer> digits)
    {

        // Make a list of items greater than 0
        ArrayList<Integer> greater0 = new ArrayList<Integer>();

        // Loop through the list of digits
        for (int digit : digits)
        {
            if (digit > 0)
            {
                greater0.add(digit);
            }
        }

        return greater0;

    }

    public static ArrayList<Integer> getSubDigits(ArrayList<Integer> digits)
    {

        // Final list of sub digits
        ArrayList<Integer> subDigits = new ArrayList<Integer>();

        // Loop through all digits but the last one
        for (int i = 0; i < digits.size() - 1; i++)
        {
            // Don't consider negative results, use absolute value
            int subDigit = Math.abs(digits.get(i) - digits.get(i + 1));

            // Add the result to the list of sub-digits
            subDigits.add(subDigit);
        }

        return subDigits;

    }
}

File: Competition/2021/Golden/Single-Digit/solution.py
# Intercollegiate Programming Competition 2021 Gold Track
# Problem 2 Single Digit

def main():
    """Run the program as a whole."""

    # Take number from STDIN
    num = input("Enter the number: ")

    # Get the list of digits from the number
    digits = split_to_digits(num)

    # List to be used in the while loop to find final single digit
    resultingList = digits[:]

    # Variable that controls if the final result is one single digit
    oneDigit = len(resultingList) == 1

    # As long as the final result is not a single digit, keep subtracting
    while not oneDigit:

        resultingList = get_sub_digits(resultingList)

        if len(only_one(resultingList)) == 1 or len(resultingList) == 1:
            oneDigit = True
    
    # Print results to STDOUT
    if only_one(resultingList):
        print(only_one(resultingList)[0])
    
    else:
        print(resultingList[0])


def split_to_digits(num):
    """Split a number into a list of its digits."""

    # Convert the number into a string
    # Loop through every single digit and add them to
    # a list as integers

    # List Comprehension in Python
    # https://www.w3schools.com/python/python_lists_comprehension.asp
    return [int(digit) for digit in str(num)]


def only_one(digits):
    """From a list of digits, check there is only one element greater than 0."""

    # Make a list of items greater than 0
    greater_0 = []

    # Loop through the list of digits
    for digit in digits:

        if digit > 0:
            greater_0.append(digit)
    
    return greater_0

def get_sub_digits(digits):
    """Return the digits resulting from subtracting a list of digits."""

    # Final list of sub digits
    sub_digits = []

    # Loop through all digits but the last one
    for i in range(len(digits) - 1):

        # Don't consider negative results
        # Instead use absolute value
        sub_digit = abs(digits[i] - digits[i + 1])
        
        # Append the result to the list of sub-digits
        sub_digits.append(sub_digit)
    
    return sub_digits

if __name__ == "__main__":
    main()